Checkov crashes when analysing google_compute_instance

**Describe the bug**
Run an analysis on this file:

```
resource "google_compute_instance" "broken" {

}
```

**To Reproduce**
Steps to reproduce the behaviour:
Place above in a file and run checkov over it.
```
jim@DESKTOP-BDL4296:/mnt/f/Code/test$ checkov -d .
ERROR:checkov.terraform.checks.resource.gcp.GoogleComputeDefaultServiceAccount:Failed to run check: Ensure that instances are not configured to use the default service account for configuration: {} at file: /google_instance.broken.tf
2021-03-09 07:26:33,669 [MainThread  ] [ERROR]  Failed to run check: Ensure that instances are not configured to use the default service account for configuration: {} at file: /google_instance.broken.tf
Traceback (most recent call last):
  File "/home/jim/.local/bin/checkov", line 5, in <module>
    run()
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/main.py", line 86, in run
    scan_reports = runner_registry.run(root_folder=root_folder, external_checks_dir=external_checks_dir,
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/common/runners/runner_registry.py", line 34, in run
    scan_report = runner.run(root_folder, external_checks_dir=external_checks_dir, files=files,
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/runner.py", line 66, in run
    self.check_tf_definition(report, root_folder, runner_filter, collect_skip_comments)
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/runner.py", line 98, in check_tf_definition
    self.run_all_blocks(definition, definitions_context, full_file_path, root_folder, report,
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/runner.py", line 105, in run_all_blocks
    self.run_block(definition[block_type], definitions_context,
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/runner.py", line 145, in run_block
    results = registry.scan(scanned_file, entity, skipped_checks, runner_filter)
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/common/checks/base_check_registry.py", line 105, in scan
    result = self.run_check(check, entity_configuration, entity_name, entity_type, scanned_file, skip_info)
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/common/checks/base_check_registry.py", line 111, in run_check
    result = check.run(scanned_file=scanned_file, entity_configuration=entity_configuration,
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/common/checks/base_check.py", line 62, in run
    raise e
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/common/checks/base_check.py", line 42, in run
    check_result['result'] = self.scan_entity_conf(entity_configuration, entity_type)
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/checks/resource/base_resource_check.py", line 17, in scan_entity_conf
    return self.scan_resource_conf(conf, entity_type)
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/checks/resource/base_resource_check.py", line 33, in wrapper
    return wrapped(self, conf)
  File "/home/jim/.local/lib/python3.8/site-packages/checkov/terraform/checks/resource/gcp/GoogleComputeDefaultServiceAccount.py", line 27, in scan_resource_conf
    if conf['name'][0].startswith('gke-'):
KeyError: 'name'
```

**Expected behaviour**
A clear and concise description of what you expected to happen.
Not to crash and raise check errors

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Desktop (please complete the following information):**
 - OS: Win10 and WSL 2
 - Checkov Version 1.0.763->1.0.832

Goes away if you suppress these rules:
  #checkov:skip=CKV_GCP_31:
  #checkov:skip=CKV_GCP_30:
  #checkov:skip=CKV_GCP_36:

--- a/checkov/kubernetes/checks/AllowedCapabilities.py
+++ b/checkov/kubernetes/checks/AllowedCapabilities.py
@@ -19,9 +19,9 @@ class AllowedCapabilities(BaseK8Check):
         return f'{conf["parent"]} - {conf["name"]}'
 
     def scan_spec_conf(self, conf):
-        if "securityContext" in conf:
-            if "capabilities" in conf["securityContext"]:
-                if "add" in conf["securityContext"]["capabilities"]:
+        if conf.get("securityContext"):
+            if conf["securityContext"].get("capabilities"):
+                if conf["securityContext"]["capabilities"].get("add"):
                     if conf["securityContext"]["capabilities"]["add"]:
                         return CheckResult.FAILED
         return CheckResult.PASSED

--- a/checkov/kubernetes/checks/AllowedCapabilitiesSysAdmin.py
+++ b/checkov/kubernetes/checks/AllowedCapabilitiesSysAdmin.py
@@ -18,9 +18,9 @@ class AllowedCapabilitiesSysAdmin(BaseK8Check):
         return f'{conf["parent"]} - {conf["name"]}'
 
     def scan_spec_conf(self, conf):
-        if "securityContext" in conf:
-            if "capabilities" in conf["securityContext"]:
-                if "add" in conf["securityContext"]["capabilities"]:
+        if conf.get("securityContext"):
+            if conf["securityContext"].get("capabilities"):
+                if conf["securityContext"]["capabilities"].get("add"):
                     if "SYS_ADMIN" in conf["securityContext"]["capabilities"]["add"]:
                         return CheckResult.FAILED
         return CheckResult.PASSED

--- a/checkov/terraform/checks/resource/gcp/GoogleComputeDefaultServiceAccount.py
+++ b/checkov/terraform/checks/resource/gcp/GoogleComputeDefaultServiceAccount.py
@@ -24,7 +24,7 @@ class GoogleComputeDefaultServiceAccount(BaseResourceCheck):
             if 'email' in conf['service_account'][0]:
                 if not re.match(DEFAULT_SERVICE_ACCOUNT, conf['service_account'][0]['email'][0]):
                     return CheckResult.PASSED
-        if conf['name'][0].startswith('gke-'):
+        if 'name' in conf and conf['name'][0].startswith('gke-'):
             return CheckResult.PASSED
         return CheckResult.FAILED
 

--- a/checkov/terraform/checks/resource/gcp/GoogleComputeDefaultServiceAccountFullAccess.py
+++ b/checkov/terraform/checks/resource/gcp/GoogleComputeDefaultServiceAccountFullAccess.py
@@ -22,7 +22,7 @@ class GoogleComputeDefaultServiceAccountFullAccess(BaseResourceCheck):
         :param conf: google_compute_instance configuration
         :return: <CheckResult>
         """
-        if conf['name'][0].startswith('gke-'):
+        if 'name' in conf and conf['name'][0].startswith('gke-'):
             return CheckResult.PASSED
         if 'service_account' in conf.keys():
             service_account_conf = conf['service_account'][0]

--- a/checkov/terraform/context_parsers/parsers/locals_context_parser.py
+++ b/checkov/terraform/context_parsers/parsers/locals_context_parser.py
@@ -10,8 +10,8 @@ class LocalsContextParser(BaseContextParser):
     def _collect_local_values(self, local_block):
         if isinstance(local_block,dict):
             for local_name, local_value in local_block.items():
-                local_value = local_value[0]
-                if type(local_value) in (int, float, bool, str):
+                local_value = local_value[0] if isinstance(local_value, list) else local_value
+                if type(local_value) in (int, float, bool, str, dict):
                     dpath.new(self.context, ['assignments', local_name], local_value)
 
     def get_block_type(self):

--- a/checkov/terraform/parser.py
+++ b/checkov/terraform/parser.py
@@ -291,7 +291,7 @@ class Parser:
             if not file_locals:
                 continue
             for k, v in file_locals[0].items():
-                locals_values[k] = v[0]
+                locals_values[k] = v[0] if isinstance(v, list) else v
 
         # Processing is done in a loop to deal with chained references and the like.
         # Loop while the data is being changed, stop when no more changes are happening.

--- a/tests/kubernetes/checks/example_AllowedCapabilities/cassandra-FAILED.yaml
+++ b/tests/kubernetes/checks/example_AllowedCapabilities/cassandra-FAILED.yaml
@@ -94,6 +94,65 @@ spec:
         volumeMounts:
         - name: cassandra-data
           mountPath: /cassandra_data
+      - name: cassandra
+        image: gcr.io/google-samples/cassandra:v13
+        imagePullPolicy: Always
+        ports:
+        - containerPort: 7000
+          name: intra-node
+        - containerPort: 7001
+          name: tls-intra-node
+        - containerPort: 7199
+          name: jmx
+        - containerPort: 9042
+          name: cql
+        resources:
+          limits:
+            cpu: "500m"
+            memory: 1Gi
+          requests:
+            cpu: "500m"
+            memory: 1Gi
+        securityContext:
+          capabilities:
+        lifecycle:
+          preStop:
+            exec:
+              command:
+              - /bin/sh
+              - -c
+              - nodetool drain
+        env:
+          - name: MAX_HEAP_SIZE
+            value: 512M
+          - name: HEAP_NEWSIZE
+            value: 100M
+          - name: CASSANDRA_SEEDS
+            value: "cassandra-0.cassandra.default.svc.cluster.local"
+          - name: CASSANDRA_CLUSTER_NAME
+            value: "K8Demo"
+          - name: CASSANDRA_DC
+            value: "DC1-K8Demo"
+          - name: CASSANDRA_RACK
+            value: "Rack1-K8Demo"
+          - name: POD_IP
+            valueFrom:
+              fieldRef:
+                fieldPath: status.podIP
+        readinessProbe:
+          exec:
+            command:
+            - /bin/bash
+            - -c
+            - /ready-probe.sh
+          initialDelaySeconds: 15
+          timeoutSeconds: 5
+        # These volume mounts are persistent. They are like inline claims,
+        # but not exactly because the names need to match exactly one of
+        # the stateful pod volumes.
+        volumeMounts:
+        - name: cassandra-data
+          mountPath: /cassandra_data
   # These are converted to volume claims by the controller
   # and mounted at the paths mentioned above.
   # do not use these in production until ssd GCEPersistentDisk or other ssd pd

--- a/tests/kubernetes/checks/test_AllowedCapabilities.py
+++ b/tests/kubernetes/checks/test_AllowedCapabilities.py
@@ -16,7 +16,7 @@ class TestAllowedCapabilities(unittest.TestCase):
         report = runner.run(root_folder=test_files_dir,runner_filter=RunnerFilter(checks=[check.id]))
         summary = report.get_summary()
 
-        self.assertEqual(summary['passed'], 2)
+        self.assertEqual(summary['passed'], 3)
         self.assertEqual(summary['failed'], 1)
         self.assertEqual(summary['skipped'], 0)
         self.assertEqual(summary['parsing_errors'], 0)

--- a/tests/terraform/checks/resource/gcp/test_GoogleComputeDefaultServiceAccountFullAccess.py
+++ b/tests/terraform/checks/resource/gcp/test_GoogleComputeDefaultServiceAccountFullAccess.py
@@ -71,6 +71,15 @@ class TestGoogleComputeDefaultServiceAccount(unittest.TestCase):
         scan_result = check.scan_resource_conf(conf=resource_conf)
         self.assertEqual(CheckResult.PASSED, scan_result)
 
+    def test_handle_empty(self):
+        hcl_res = hcl2.loads("""
+        resource "google_compute_instance" "broken" {
+        }
+        """)
+        resource_conf = hcl_res['resource'][0]['google_compute_instance']['broken']
+        scan_result = check.scan_resource_conf(conf=resource_conf)
+        self.assertEqual(CheckResult.PASSED, scan_result)
+
 
 if __name__ == '__main__':
     unittest.main()

--- a/tests/terraform/context_parsers/test_locals_parser.py
+++ b/tests/terraform/context_parsers/test_locals_parser.py
@@ -1,0 +1,44 @@
+import unittest
+from checkov.terraform.parser import Parser
+from checkov.terraform.context_parsers.registry import parser_registry
+import os
+
+
+class TestLocalsContextParser(unittest.TestCase):
+
+    def setup_dir(self, rel_path):
+        test_root_dir = os.path.dirname(os.path.realpath(__file__)) + rel_path
+        tf_definitions = {}
+        parsing_errors = {}
+        definitions_context = {}
+        Parser().parse_directory(directory=test_root_dir,
+                                 out_definitions=tf_definitions,
+                                 out_parsing_errors=parsing_errors)
+        for definition in tf_definitions.items():
+            definitions_context = parser_registry.enrich_definitions_context(definition)
+        return definitions_context
+
+    def test_assignments_exists(self):
+        definitions_context = self.setup_dir('/../evaluation/resources/default_evaluation/')
+        assignments = definitions_context[os.path.dirname(os.path.realpath(__file__)) + '/../evaluation/resources/default_evaluation/main.tf']['locals']['assignments']
+        self.assertIsNotNone(assignments)
+
+        expected_assignments = {'dummy_with_dash': '${format("-%s",dummy_1)}', 'dummy_with_comma': '${format(":%s",dummy_1)}', 'bucket_name': 'MyBucket'}
+
+        for k, v in assignments.items():
+            self.assertEqual(expected_assignments[k], v)
+
+    def test_assignment_value(self):
+        definitions_context = self.setup_dir('/../evaluation/resources/locals_evaluation/')
+        assignments = definitions_context[
+            os.path.dirname(os.path.realpath(__file__)) + '/../evaluation/resources/locals_evaluation/main.tf'][
+            'locals'].get('assignments')
+        self.assertIsNotNone(assignments)
+        self.assertEqual(1, len(assignments.items()))
+        for k, v in assignments.items():
+            self.assertEqual(k, 'common_tags')
+            self.assertIsInstance(v, dict)
+
+
+if __name__ == '__main__':
+    unittest.main()

--- a/tests/terraform/evaluation/resources/locals_evaluation/main.tf
+++ b/tests/terraform/evaluation/resources/locals_evaluation/main.tf
@@ -1,0 +1,5 @@
+locals = {
+  common_tags = {
+    Env = "dev"
+  }
+}

